Pioneer: Runtime upgrade proposal optimalization

Lezek123 opened the pull request:
This PR fixes the problem with major lags on `RuntimeUpgrade` proposal details page which was raised in issue https://github.com/Joystream/joystream/issues/651

The fix also benefits proposal module in general, as it caches details for the `RuntimeUpgrade` proposal throught routes, which helps us avoid repeating a few time-consuming/computation-heavy tasks like fetching `RuntimeUpgrade` proposal details or re-calculating the WASM bytecode hash.

Further optimalizations are possible, but this already reduces the loading time greatly (especially when moving from one proposal-related page to another), gets rid of post-loading lags and seems like a good, quick optimalization in general.

Commit:
Runtime upgrade proposal optimalization

Commit:
Linting

mnaamani commented on pioneer/packages/joy-utils/src/transport/proposals.ts:
```diff
@@ -61,10 +64,44 @@ export default class ProposalsTransport extends BaseTransport {
     return (this.api.consts.proposalsEngine.cancellationFee as BalanceOf).toNumber();
   }
 
+  private runtimeUpgradeProposalCachedDetails (id: ProposalId) {
```
The name of the method `runtimeUpgradeProposalCachedDetails` and the private property `runtimeUpgradeProposalDetailsCache` are too similar and confused me a but while reading the code.
perhaps rename the method:
 `getRuntimeUpgradeProposalCachedDetails`

mnaamani commented on pioneer/packages/joy-utils/src/transport/proposals.ts:
```diff
@@ -61,10 +64,44 @@ export default class ProposalsTransport extends BaseTransport {
+  private runtimeUpgradeProposalDetails (id: ProposalId, wasm: Bytes): [string, number] {
+    const cachedDetails = this.runtimeUpgradeProposalCachedDetails(id);
+
+    if (cachedDetails) {
```
this will never be truthy because it is only called by `proposalById()` on cache miss.

It also suggests the name of the method should be `updateRuntimeUpgradeProposalDetails` or `computeAndCacheRuntimeUpgradeProposalDetails`

mnaamani commented:
Tested works well, just had some suggestions.

There is a merge conflict to fix.

Also I'm wonder if you considered using the `APIQueryCache` as you did in the joy-roles app. Would it help or make things worse in caching the WASM blob in memory?

mnaamani commented:
Does it make sense to make a similar patch for master branch to benefit from this before the release?

Commit:
Merge branch 'nicaea' into pioneer-proposals-optimalization (+conflict fix)

Lezek123 commented:
> Also I'm wonder if you considered using the APIQueryCache as you did in the joy-roles app. Would it help or make things worse in caching the WASM blob in memory?

`APIQueryCache` only caches the result of given query at given block and is reset every time there's a new block, so we can use it as an additional optimalization for the transports in case some component gets reloaded and for some reason tries to re-execute the samie queries that were already executed at given block.

> Does it make sense to make a similar patch for master branch to benefit from this before the release?

I think it does! Will prepare another PR to master

Commit:
Add query cache and fix "cache throught routes" bug

Commit:
Cache details for all proposal types + some cleanup

Lezek123 commented:
- Fixed conflicts
- Added query cache for `joy-utils` transports
- Fixed a bug with transport instance beeing re-constructed throught proposal routes (which ie. prevented caching details when moving between `/proposals` and `/proposals/ID` routes)
- Implemented caching for all proposal details (not just runtime upgrade)
- Made the proposal details caching code cleaner

mnaamani commented on pioneer/packages/joy-utils/src/transport/proposals.ts:
```diff
@@ -112,13 +149,9 @@ export default class ProposalsTransport extends BaseTransport {
-    const vote = await this.proposalsEngine.voteExistsByProposalByVoter<VoteKind>(proposalId, voterId);
-    const hasVoted = (await this.proposalsEngine.voteExistsByProposalByVoter.size(proposalId, voterId)).toNumber();
+    const vote = (await this.proposalsEngine.voteExistsByProposalByVoter(proposalId, voterId)) as VoteKind;
+    const hasVoted = (await this.api.query.proposalsEngine.voteExistsByProposalByVoter.size(proposalId, voterId)).toNumber();
```
I see you are bypassing the cached api here, why exactly?

Lezek123 replied:
Cached api doesn't have `.size()` available

mnaamani commented on pioneer/packages/joy-utils/src/transport/council.ts:
```diff
@@ -8,21 +8,22 @@ import { FIRST_MEMBER_ID } from '../consts/members';
     if (atBlock) {
       const blockHash = await this.chainT.blockHash(atBlock);
-      return ((await this.council.activeCouncil.at(blockHash)) as Seats).length;
+      return ((await this.api.query.council.activeCouncil.at(blockHash)) as Seats).length;
```
bypassing the cached api? Any special reason?

Lezek123 replied:
Cached api doesn't have `.at()` available

mnaamani commented on pioneer/packages/joy-utils/src/transport/proposals.ts:
```diff
@@ -177,7 +210,7 @@ export default class ProposalsTransport extends BaseTransport {
 
   async subscribeProposal (id: number|ProposalId, callback: () => void) {
-    return this.proposalsEngine.proposals(id, callback);
+    return this.api.query.proposalsEngine.proposals(id, callback);
```
this one here I understand because you want to subsribe to updates..?

Lezek123 replied:
Yes, this is also not available via cached api. Only simple queries are supported.

mnaamani replied:
ok thanks for clarification

mnaamani reviewed (approved):
Looks good, just had a query on why you were bypassing the cached api in a few distinct places.
Tested against staging network works great
